### PR TITLE
Apply annotations dependency to `testFixturesImplementation` 

### LINKED ISSUE #393: Annotation not available to testFixtures post-Kotlin 2.0.10
Kotlin 2.0.10 [fixed a bug around friendPaths and test-fixtures](https://youtrack.jetbrains.com/issue/KT-70525/Unresolved-reference-caused-by-testFixtures#focus=Comments-27-10333047.0-0), where previously `implementation` dependencies would be exposed to the fixtures compile classpath. 

Using `@Poko` in main sources without adding `testFixturesImplementation("dev.drewhamilton.poko:poko-annotations:0.17.0")` results in a compilation error due to missing classes:

```
> Task :my-module:compileTestFixturesKotlin FAILED
e: Could not find class <dev/drewhamilton/poko/Poko>
```

A thought to address this: Configuring the annotations dependency as `api` doesn't feel like the right solution. Is it possible for the Poko Gradle Plugin to add this dependency to the test fixtures configuration if java-test-fixtures is applied?

### PR DESCRIPTION
**Pull request by jamiesanson**
When applying Poko to JVM projects with the `java-test-fixtures` Gradle Plugin applied, add `poko-annotations` to `testFixturesImplementation`.

Closes #393.

### CODE REVIEW

**Comment by JakeWharton** [comment]
I think I would prefer this still be a warning. Kotlin 2.1.0 will allow suppressing specific warnings by name even when using the warnings-as-error setting (see https://youtrack.jetbrains.com/issue/KT-8087).

We'll see what @drewhamilton thinks.

**Comment by drewhamilton** [comment]
The concern is that a library author adds `com.example.MyData` to their classpath, but then typos something like `pokoAnnotation = "comm/example/MyData"` in their configuration, so they have a library full of `@MyData` annotations but no code generation. This scenario should remain at least a warning, but I'm not convinced it shouldn't be an error.

In #393, Jake said:
> It might be better to no-op the compiler plugin if the runtime isn't present in a compilation unit.

Why is this better?

**Comment by JakeWharton** [comment]
The motivating example was the compiler plugin being applied to all Kotlin compilations in the Gradle project (including test fixtures, specifically) but the annotation not being present in the associated configurations.

**Comment by drewhamilton** [comment]
Yes, sorry, I meant why is this approach better than either (a) not applying Poko to test fixture configurations by default, or (b) adding the annotations dep to the test fixtures dependency by default?

My gut says (a) makes more sense, because maintaining binary compatibility in test fixtures is rarely if ever needed, but maybe that's wrong?

**Comment by JakeWharton** [comment]
I don't believe we get a choice in the fact that the compiler plugin is applied. Test fixtures support compiling Kotlin and I would expect any plugins to go along with that automatically (with our use of the subplugin plugin thing).

We can add the first-party dep automatically by special-casing the test fixture plugin. That was proposed. The only case you'd need to manually intervene then is if you did a compileOnly of your own dep with an annotation.

**Comment by jamiesanson** [comment]
Yeah, this PR isn't ideal - hadn't properly considered that typo case, agree this should be at minimum a warning.

---

My original suggestion was to special case it, and add the annotations dependency automatically. Test fixtures end up being published artifacts of a library, so figured that while maintaining binary compatibility probably isn't needed _all_ the time, some library authors may strive for it. Having Poko automatically apply this dependency to testFixture compilations as it does to other ones feels preferable here. 

Manual intervention would be pretty obviously needed if a developer typos, or adds their own annotation dependency to the wrong configuration - this error would persist, and could be made a little more descriptive to point them in the right direction.

**Comment by drewhamilton** [comment]
> I don't believe we get a choice in the fact that the compiler plugin is applied. Test fixtures support compiling Kotlin and I would expect any plugins to go along with that automatically (with our use of the subplugin plugin thing).

`KotlinCompilerPluginSupportPlugin` has an `isApplicable` function that we set to `true` but could be more selective based on the input `KotlinCompilation`. So I think we could skip applying it to test fixtures if we wanted to.

But if test fixtures are generally published with the library, it seems like special-casing the inclusion of the annotation dep for test fixtures, as @jamiesanson initially proposed, is the most intuitive behavior.

File: poko-gradle-plugin/src/main/kotlin/dev/drewhamilton/poko/gradle/PokoGradlePlugin.kt [hunk 1]
```diff
@@ -30,6 +30,9 @@ public class PokoGradlePlugin : KotlinCompilerPluginSupportPlugin {
                         }
                     }
                 } else {
+                    if (target.plugins.hasPlugin("org.gradle.java-test-fixtures")) {
+                        target.dependencies.add("testFixturesImplementation", annotationDependency)
+                    }
```
Comment (JakeWharton) [inline poko-gradle-plugin/src/main/kotlin/dev/drewhamilton/poko/gradle/PokoGradlePlugin.kt:35]: This is probably fine to start with. We may have to move this outside the `if` at some point as support is expanded.

**Review by JakeWharton — APPROVED** [review]

**Comment by drewhamilton** [comment]
Sorry for the delay – looks great, and thanks for adding to the sample too.

**Review by drewhamilton — APPROVED** [review]

**Pull request merged**